### ISSUE TITLE
批量写入测试,速度好像没有jdbc driver的快

### ISSUE THREAD

**Issue by gj-zhang** [housepower/ClickHouse-Native-JDBC#245]
### Environment

- OS version: CentOS Linux release 7.6.1810 (Core)
- JDK version: java version "1.8.0_102"
- ClickHouse Server version: 20.9.3.45 on docker 
- ClickHouse Native JDBC version: 2.5.0-SNAPSHOT
- (Optional) Spark version: N/A
- (Optional) Other components' version:  flume 1.9.0
- 硬件  ssd  40c 256G

### 场景
flume和clickhouse部署在同一台机器,  使用flume 消费kafka数据写入clickhouse 本地表, 自己写了flume sink, 大概打印了一下每个batch写入20w和50w的场景耗时发现jdbc driver更快, 连接池连接数100, 麻烦帮忙看下是不是我使用的有问题还是其他问题
## 表结构
```
CREATE TABLE default.store_sale_tcp_test
(
    `ss_sold_date_sk` String,
    `ss_sold_time_sk` String,
    `ss_item_sk` String,
    `ss_customer_sk` String,
    `ss_cdemo_sk` String,
    `ss_hdemo_sk` String,
    `ss_addr_sk` String,
    `ss_store_sk` String,
    `ss_promo_sk` String,
    `ss_ticket_number` String,
    `ss_quantity` String,
    `ss_wholesale_cost` String,
    `ss_list_price` String,
    `ss_sales_price` String,
    `ss_ext_discount_amt` String,
    `ss_ext_sales_price` String,
    `ss_ext_wholesale_cost` String,
    `ss_ext_list_price` String,
    `ss_ext_tax` String,
    `ss_coupon_amt` String,
    `ss_net_paid` String,
    `ss_net_paid_inc_tax` String,
    `ss_net_profit` String,
    `app_id` String,
    `module` String,
    `class_name` String,
    `log_msg` String
)
ENGINE = ReplicatedMergeTree('/clickhouse/tables/{cluster}/test/store_sale_tcp_test/{shard}', '{replica}')
ORDER BY ss_sold_date_sk
SETTINGS index_granularity = 8192
```
日志只截取了部分， 测试数据是tpc-ds 造出来的store_sales表的数据
### jdbc 日志打印 处理50w行, 时间2000多毫秒
```
10 Dec 2020 15:41:58,258 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2084
10 Dec 2020 15:42:06,734 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2175
10 Dec 2020 15:42:14,851 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2137
10 Dec 2020 15:42:24,871 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2091
10 Dec 2020 15:42:33,121 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2202
10 Dec 2020 15:42:41,421 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2089
10 Dec 2020 15:42:49,959 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2076
10 Dec 2020 15:42:57,960 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2005
10 Dec 2020 15:43:06,343 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 500000, time 2091

```
### native jdbc 日志打印 处理50w行, 时间3000多毫秒到4000多毫秒
```
10 Dec 2020 11:35:21,009 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 4333
10 Dec 2020 11:35:30,727 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3154
10 Dec 2020 11:35:40,384 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3116
10 Dec 2020 11:35:49,917 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3141
10 Dec 2020 11:35:59,547 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3181
10 Dec 2020 11:36:09,071 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3209
10 Dec 2020 11:36:19,845 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3101
10 Dec 2020 11:36:29,149 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3099
10 Dec 2020 11:36:38,228 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3145
10 Dec 2020 11:36:47,765 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3109
10 Dec 2020 11:36:56,583 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3057
10 Dec 2020 11:37:07,574 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 500000, time 3170
```
### jdbc 日志打印 处理20w行 时间 800多毫秒
```
10 Dec 2020 16:04:45,764 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 845
10 Dec 2020 16:04:49,050 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 802
10 Dec 2020 16:04:52,537 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 829
10 Dec 2020 16:04:55,930 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 831
10 Dec 2020 16:04:59,178 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 804
10 Dec 2020 16:05:02,589 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 807
10 Dec 2020 16:05:06,037 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 838
10 Dec 2020 16:05:09,211 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 808
10 Dec 2020 16:05:12,545 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 766
10 Dec 2020 16:05:16,040 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 809
10 Dec 2020 16:05:19,461 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 823
10 Dec 2020 16:05:22,627 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 830
10 Dec 2020 16:05:25,885 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseSink2.process:187)  - handle 200000, time 849
```
### native jdbc 日志打印 处理20w行
```
10 Dec 2020 16:13:06,525 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1191
10 Dec 2020 16:13:10,518 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1163
10 Dec 2020 16:13:14,275 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1135
10 Dec 2020 16:13:17,861 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1161
10 Dec 2020 16:13:21,675 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1154
10 Dec 2020 16:13:25,454 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1142
10 Dec 2020 16:13:28,982 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1126
10 Dec 2020 16:13:32,706 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1139
10 Dec 2020 16:13:38,151 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1166
10 Dec 2020 16:13:41,739 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1140
10 Dec 2020 16:13:45,499 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1146
10 Dec 2020 16:13:49,077 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1177
10 Dec 2020 16:13:52,681 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1129
10 Dec 2020 16:13:56,394 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1139
10 Dec 2020 16:13:59,864 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1144
10 Dec 2020 16:14:03,558 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (com.xxx.flume.ck.FlumeClickHouseTcpSink.process:177)  - handle 200000, time 1138
```

### jdbc 代码
``` Java
public class FlumeClickHouseSink2 extends AbstractSink implements Configurable {

    private static final Logger logger = LoggerFactory.getLogger(FlumeClickHouseSink2.class);

    private BalancedClickhouseDataSource dataSource = null;
    private HikariDataSource ds = null;
    private SinkCounter sinkCounter = null;
    private String host = null;
    private String port = null;
    private String user = null;
    private String password = null;
    private String database = null;
    private String table = null;

    private int batchSize;
    private ClickHouseFormat format;
    private String[] appIds = null;
    private String[] moduleNames = null;
    private String[] classNames = null;
    private String[] msgs = null;
    // initAppId initModule initClassName initMsg 写死了几万条数据随机选择作为表最后四个字段的值
    private void initAppId() {
        this.appIds = new String[]{};
    }

    private void initModule() {
        this.moduleNames = new String[]{};
    }

    private void initClassName() {
        this.classNames = new String[]{};
    }

    private void initMsg() {
        this.msgs = new String[]{};
    }

    @Override
    public void configure(Context context) {
        logger.info("ClickHouse start configure...");

        if (sinkCounter == null) {
            sinkCounter = new SinkCounter(getName());
        }

        this.host = context.getString(HOST);
        if (!this.host.startsWith("jdbc:clickhouse://")) {
            this.host = "jdbc:clickhouse://" + this.host;
        }
        this.database = context.getString(DATABASE);
        this.table = context.getString(TABLE);
        this.port = context.getString(PORT, DEFAULT_PORT);
        this.user = context.getString(USER, DEFAULT_USER);
        this.password = context.getString(PASSWORD, DEFAULT_PASSWORD);
        this.batchSize = context.getInteger(BATCH_SIZE, DEFAULT_BATCH_SIZE);
        initAppId();
        initClassName();
        initModule();
        initMsg();
    }

    @Override
    public void start() {
        logger.info("ClickHouse sink {} starting", getName());
        String jdbcUrl = String.format("%s:%s/%s", this.host, this.port, this.database);
//        ClickHouseProperties properties = new ClickHouseProperties().withCredentials(this.user, this.password);
        logger.info("ClickHouse url: {}, user: {}, password: {}", jdbcUrl, user, password);
//        this.dataSource = new BalancedClickhouseDataSource(jdbcUrl, properties);
        HikariConfig config = new HikariConfig();
        config.setJdbcUrl(jdbcUrl);
        config.setUsername(this.user);
        config.setPassword(this.password);
        config.setMaximumPoolSize(100);
        config.setDriverClassName("ru.yandex.clickhouse.ClickHouseDriver");
        this.ds = new HikariDataSource(config);
        sinkCounter.start();
        super.start();
        logger.info("ClickHouse DataSource start");
    }


    @Override
    public void stop() {
        logger.info("ClickHouse sink {} stopping", getName());
        sinkCounter.incrementConnectionClosedCount();
        sinkCounter.stop();
        super.stop();
    }


    @Override
    public Status process() throws EventDeliveryException {
        Status status = null;

        // Start transaction
        Channel ch = getChannel();
        Transaction txn = ch.getTransaction();
        txn.begin();
        StringBuilder batch = new StringBuilder();
        try {
            Connection conn = ds.getConnection();
            int count;
            String sql = "insert  into " + database + "." + table
                    + "(ss_sold_date_sk,ss_sold_time_sk,ss_item_sk,ss_customer_sk,ss_cdemo_sk,ss_hdemo_sk,ss_addr_sk,ss_store_sk,ss_promo_sk,ss_ticket_number,ss_quantity,ss_wholesale_cost,ss_list_price,ss_sales_price,ss_ext_discount_amt,ss_ext_sales_price,ss_ext_wholesale_cost,ss_ext_list_price,ss_ext_tax,ss_coupon_amt,ss_net_paid,ss_net_paid_inc_tax,ss_net_profit, app_id, module, class_name, log_msg) "
                    + "values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?);";
            PreparedStatement preparedStatement = conn.prepareStatement(sql);
            int nullCount = 0;
            for (count = 0; count < batchSize; ++count) {
                Event event = ch.take();
                if (event == null) {
                    nullCount++;
                    if (nullCount % 100 == 0) {
                        logger.info("null count: {}", nullCount);
                    }
                    continue;
                }
                String s = new String(event.getBody(), "UTF-8");
                StoreSales storeSales = JSON.parseObject(s, StoreSales.class);
                preparedStatement.setString(1, storeSales.getSs_sold_date_sk());
                preparedStatement.setString(2, storeSales.getSs_sold_time_sk());
                preparedStatement.setString(3, storeSales.getSs_item_sk());
                preparedStatement.setString(4, storeSales.getSs_customer_sk());
                preparedStatement.setString(5, storeSales.getSs_cdemo_sk());
                preparedStatement.setString(6, storeSales.getSs_hdemo_sk());
                preparedStatement.setString(7, storeSales.getSs_addr_sk());
                preparedStatement.setString(8, storeSales.getSs_store_sk());
                preparedStatement.setString(9, storeSales.getSs_promo_sk());
                preparedStatement.setString(10, storeSales.getSs_ticket_number());
                preparedStatement.setString(11, storeSales.getSs_quantity());
                preparedStatement.setString(12, storeSales.getSs_wholesale_cost());
                preparedStatement.setString(13, storeSales.getSs_list_price());
                preparedStatement.setString(14, storeSales.getSs_sales_price());
                preparedStatement.setString(15, storeSales.getSs_ext_discount_amt());
                preparedStatement.setString(16, storeSales.getSs_ext_sales_price());
                preparedStatement.setString(17, storeSales.getSs_ext_wholesale_cost());
                preparedStatement.setString(18, storeSales.getSs_ext_list_price());
                preparedStatement.setString(19, storeSales.getSs_ext_tax());
                preparedStatement.setString(20, storeSales.getSs_coupon_amt());
                preparedStatement.setString(21, storeSales.getSs_net_paid());
                preparedStatement.setString(22, storeSales.getSs_net_paid_inc_tax());
                preparedStatement.setString(23, storeSales.getSs_net_profit());
                String appId = appIds[count % 52];
                String moduleName = moduleNames[count % 26];
                String className = classNames[count % 26];
                preparedStatement.setString(24, appId);
                preparedStatement.setString(25, moduleName);
                preparedStatement.setString(26, className);
                preparedStatement.setString(27, appId + moduleName + className + "--" + msgs[count % 1000]);

                preparedStatement.addBatch();
            }
            if (count <= 0) {
                sinkCounter.incrementBatchEmptyCount();
                txn.commit();
                return Status.BACKOFF;
            } else if (count < batchSize) {
                sinkCounter.incrementBatchUnderflowCount();
            } else {
                sinkCounter.incrementBatchCompleteCount();
            }
            sinkCounter.addToEventDrainAttemptCount(count);
            long start = System.currentTimeMillis();
            preparedStatement.executeBatch();
            // 统计值
            logger.info("handle {}, time {}", count, System.currentTimeMillis() - start);
            sinkCounter.incrementEventDrainSuccessCount();
            status = Status.READY;
            txn.commit();
        } catch (Throwable t) {
            txn.rollback();
            logger.error(t.getMessage(), t);
            logger.error(" ClickHouse raw str: {}", batch.toString());
            status = Status.BACKOFF;
            // re-throw all Errors
            if (t instanceof Error) {
                throw (Error) t;
            }
        } finally {
            txn.close();
        }
        return status;
    }
}
```

### native jdbc 代码
```Java
public class FlumeClickHouseTcpSink extends AbstractSink implements Configurable {

    private static final Logger logger = LoggerFactory.getLogger(FlumeClickHouseTcpSink.class);

    private HikariDataSource dataSource = null;
    private SinkCounter sinkCounter = null;
    private String host = null;
    private String port = null;
    private String user = null;
    private String password = null;
    private String database = null;
    private String table = null;

    private int connectionCount = 1;
    private int batchSize;
    private String[] appIds = null;
    private String[] moduleNames = null;
    private String[] classNames = null;
    private String[] msgs = null;

    private void initAppId() {
        this.appIds = new String[]{};
    }

    private void initModule() {
        this.moduleNames = new String[]{};
    }

    private void initClassName() {
        this.classNames = new String[]{};
    }

    private void initMsg() {
        this.msgs = new String[]{};
    }

    @Override
    public void configure(Context context) {
        logger.info("ClickHouse start configure...");

        if (sinkCounter == null) {
            sinkCounter = new SinkCounter(getName());
        }

        this.host = context.getString(HOST);
        if (!this.host.startsWith("jdbc:clickhouse://")) {
            this.host = "jdbc:clickhouse://" + this.host;
        }
        this.database = context.getString(DATABASE);
        this.table = context.getString(TABLE);
        this.port = context.getString(PORT, DEFAULT_PORT);
        this.user = context.getString(USER, DEFAULT_USER);
        this.password = context.getString(PASSWORD, DEFAULT_PASSWORD);
        this.batchSize = context.getInteger(BATCH_SIZE, DEFAULT_BATCH_SIZE);
        this.connectionCount = context.getInteger(CONNECTION_COUNT, DEFAULT_CONNECTION_COUNT);
        initAppId();
        initClassName();
        initModule();
        initMsg();
    }

    @Override
    public void start() {
        logger.info("ClickHouse sink {} starting", getName());
        String jdbcUrl = String.format("%s:%s/%s", this.host, this.port, this.database);
        HikariConfig conf = new HikariConfig();
        conf.setJdbcUrl(jdbcUrl);
        conf.setUsername(this.user);
        conf.setPassword(this.password);
        conf.setDriverClassName("com.github.housepower.jdbc.ClickHouseDriver");
        conf.setMaximumPoolSize(this.connectionCount);
        logger.info("ClickHouse url: {}, user: {}, password: {}", jdbcUrl, user, password);
        dataSource = new HikariDataSource(conf);
        sinkCounter.start();
        super.start();
        logger.info("ClickHouse DataSource start");
    }


    @Override
    public void stop() {
        logger.info("ClickHouse sink {} stopping", getName());
        sinkCounter.incrementConnectionClosedCount();
        sinkCounter.stop();
        dataSource.close();
        super.stop();
    }


    @Override
    public Status process() throws EventDeliveryException {
        Status status = null;

        // Start transaction
        Channel ch = getChannel();
        Transaction txn = ch.getTransaction();
        txn.begin();
        StringBuilder batch = new StringBuilder();
        try {
            Connection conn = dataSource.getConnection();
            int count;
            String sql = "insert  into " + database + "." + table
                    + "(ss_sold_date_sk,ss_sold_time_sk,ss_item_sk,ss_customer_sk,ss_cdemo_sk,ss_hdemo_sk,ss_addr_sk,ss_store_sk,ss_promo_sk,ss_ticket_number,ss_quantity,ss_wholesale_cost,ss_list_price,ss_sales_price,ss_ext_discount_amt,ss_ext_sales_price,ss_ext_wholesale_cost,ss_ext_list_price,ss_ext_tax,ss_coupon_amt,ss_net_paid,ss_net_paid_inc_tax,ss_net_profit, app_id, module, class_name, log_msg) "
                    + "values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?);";
            PreparedStatement preparedStatement = conn.prepareStatement(sql);
            int nullCount = 0;
            for (count = 0; count < batchSize; ++count) {
                Event event = ch.take();
                if (event == null) {
                    nullCount++;
                    if (nullCount % 100 == 0) {
                        logger.info("null count: {}", nullCount);
                    }
                    continue;
                }
                String s = new String(event.getBody(), "UTF-8");
                StoreSales storeSales = JSON.parseObject(s, StoreSales.class);
                preparedStatement.setString(1, storeSales.getSs_sold_date_sk());
                preparedStatement.setString(2, storeSales.getSs_sold_time_sk());
                preparedStatement.setString(3, storeSales.getSs_item_sk());
                preparedStatement.setString(4, storeSales.getSs_customer_sk());
                preparedStatement.setString(5, storeSales.getSs_cdemo_sk());
                preparedStatement.setString(6, storeSales.getSs_hdemo_sk());
                preparedStatement.setString(7, storeSales.getSs_addr_sk());
                preparedStatement.setString(8, storeSales.getSs_store_sk());
                preparedStatement.setString(9, storeSales.getSs_promo_sk());
                preparedStatement.setString(10, storeSales.getSs_ticket_number());
                preparedStatement.setString(11, storeSales.getSs_quantity());
                preparedStatement.setString(12, storeSales.getSs_wholesale_cost());
                preparedStatement.setString(13, storeSales.getSs_list_price());
                preparedStatement.setString(14, storeSales.getSs_sales_price());
                preparedStatement.setString(15, storeSales.getSs_ext_discount_amt());
                preparedStatement.setString(16, storeSales.getSs_ext_sales_price());
                preparedStatement.setString(17, storeSales.getSs_ext_wholesale_cost());
                preparedStatement.setString(18, storeSales.getSs_ext_list_price());
                preparedStatement.setString(19, storeSales.getSs_ext_tax());
                preparedStatement.setString(20, storeSales.getSs_coupon_amt());
                preparedStatement.setString(21, storeSales.getSs_net_paid());
                preparedStatement.setString(22, storeSales.getSs_net_paid_inc_tax());
                preparedStatement.setString(23, storeSales.getSs_net_profit());
                String appId = appIds[count % 52];
                String moduleName = moduleNames[count % 26];
                String className = classNames[count % 26];
                preparedStatement.setString(24, appId);
                preparedStatement.setString(25, moduleName);
                preparedStatement.setString(26, className);
                preparedStatement.setString(27, appId + moduleName + className + "--" + msgs[count % 1000]);

                preparedStatement.addBatch();
            }
            if (count <= 0) {
                sinkCounter.incrementBatchEmptyCount();
                txn.commit();
                return Status.BACKOFF;
            } else if (count < batchSize) {
                sinkCounter.incrementBatchUnderflowCount();
            } else {
                sinkCounter.incrementBatchCompleteCount();
            }
            sinkCounter.addToEventDrainAttemptCount(count);
            long start = System.currentTimeMillis();
            int[] ints = preparedStatement.executeBatch();
            logger.info("handle {}, time {}", count, System.currentTimeMillis() - start);
            sinkCounter.incrementEventDrainSuccessCount();
            status = Status.READY;
            txn.commit();
        } catch (Throwable t) {
            txn.rollback();
            logger.error(t.getMessage(), t);
            logger.error(" ClickHouse raw str: {}", batch.toString());
            status = Status.BACKOFF;
            // re-throw all Errors
            if (t instanceof Error) {
                throw (Error) t;
            }
        } finally {
            txn.close();
        }
        return status;
    }
}
```

**Comment by gj-zhang** [comment]
anybody help me?

**Comment by pan3793** [comment]
TL:DR Currently, this native driver has no advantage performance over yandex jdbc driver on `String` type.

Basically, the differences between this native driver and yandex jdbc driver are:
- tcp vs http, because http also support keep alive, there is not much overhead
- this native client parse sql in client side, and associate data by column with zlib compress; yandex jdbc driver just send sql literal to clickhouse driver also compressed, so if all columns are `String` type, we have no advantage but some overhead due to memory copy. (Currently memory management is not good)

Further, clickhouse doesn't support transaction, and `txn.rollback()` will do nothing. 

> 连接池连接数100

We usually set 1 or 2 per node.

**Comment by sundy-li** [comment]
[这里](https://github.com/housepower/ClickHouse-Native-JDBC/runs/1523501180
)有 String 的测试对比用例, 相比起来性能差不多； 

不知道你的batch是多大

```
Benchmark                                           (batchSize)  (columnNum)  (selectNumber)  Mode  Cnt      Score   Error  Units
WideColumnStringInsertIBenchmark.benchInsertHttp         200000           20             N/A  avgt         468.486          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         200000           50             N/A  avgt        1186.521          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         500000           20             N/A  avgt        1216.193          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         500000           50             N/A  avgt        2948.550          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       200000           20             N/A  avgt         459.299          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       200000           50             N/A  avgt        1187.662          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       500000           20             N/A  avgt        1127.769          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       500000           50             N/A  avgt        3040.152          ms/op
```

**Comment by gj-zhang** [comment]
ok thank you

**Comment by gj-zhang** [comment]
> TL;DR Currently, this native driver has no advantage performance over yandex jdbc driver on `String` type.
> 
> Basically, the differences between this native driver and yandex jdbc driver are:
> 
> * tcp vs http, because http also support keep alive, there is not much overhead
> * this native client parse sql in client side, and associate data by column with zlib compress; yandex jdbc driver just send sql literal to clickhouse server also in compressed, so if all columns are `String` type, we have no advantage but some overhead due to memory copy. (Currently memory management is not good)
> 
> Further, clickhouse doesn't support transaction, and `txn.rollback()` will do nothing.
> 
> > 连接池连接数100
> 
> We usually set 1 or 2 connections per node per table in batch insert scenario.
